core/src/main/scala/cromwell/core/simpleton/WomValueBuilder.scala

```diff
@@ -85,36 +99,68 @@ object WomValueBuilder {
         case MapElementPattern("right", more) => PairRight -> component.copy(path = more)
       }
     }
-
-    // Group tuples by key using a Map with key type `K`.
-    def group[K](tuples: Traversable[(K, SimpletonComponent)]): Map[K, Traversable[SimpletonComponent]] = {
-      tuples groupBy { case (i, _) => i } mapValues { _ map { case (_, s) => s} }
+    
+    def toWomFile(components: Traversable[SimpletonComponent]) = {
+      // If there's just one simpleton, it's a single file
+      if (components.size == 1) toWomValue(WomSingleFileType, components)
```

Comment:
Is this always true?

`WomUnlistedDirectoryType` is also a `WomPrimitiveType`, matching on the `case _: WomPrimitiveType =>` below.

Reply:
I'm still digesting all this, so happy to discuss further.

There used to only be _one_ `WomType` for _all_ files, the `WomFileType`. However, that type could be two different `WomValue` instances, either a `WomSingleFile` or a `WomGlobFile`. Now there is no concrete `WomFileType`. It's a trait instead, like `WomPrimitiveType`. The old type with a display string "File" is mapped to `WomSingleFileType`.

Before:
https://github.com/broadinstitute/cromwell/blob/cbdc3bc80736ae2ae007c8956c50d7db626af07e/wom/src/main/scala/wom/types/WomFileType.scala#L8-L9

Currently:
https://github.com/broadinstitute/cromwell/blob/063a4516487c21dc9e96292f0cf2786b15b584e1/wom/src/main/scala/wom/types/WomFileType.scala#L8

I'm wonder if...

For each of the three `WomPrimitive` types:
- `WomGlobFile`
- `WomSingleFile`
- `WomUnlistedDirectory`

...use existing code for singletons that writes the type + the-single-value (at least I _think_ it writes the types).

For the other non-primitive `WomFile`:
- `WomMaybeListedDirectory`
- `WomMaybePopulatedFile`

...write them as one would a recursive `WomMap`/`WomObject`. This shouldn't require a "marker" anymore, assuming we know via the type that instead the rows should be converted to a `WomMaybeListedDirectory`/`WomMaybePopulatedFile`?

I'm also realizing that the `toDisplayString` may be what is stored in the database and should be added to the non-primitive types. However, I also don't know how much that "display string" in WOM must be the same as it is / will be defined in the WDL spec.

Reply:
Sounds good for the primitive `WomFileType`s
However 
> assuming we know via the type that instead the rows should be converted to a WomMaybeListedDirectory/WomMaybePopulatedFile

I don't believe we do. For example `WomMaybeListedDirectory` has a `listingOption: Option[Seq[WomFile]]` attribute.
When we simpletonize those files we lose the information of what type of file exactly was it (specifically was it a WomMaybeListedDirectory or a WomMaybePopulatedFile ? - because they're both going to end up being simpletonized like objects). This is what the marker is trying to make up for.
When we want to build them back to `WomValue` the only thing we know is that they are `WomFile`s.